### PR TITLE
add mypy check on CI

### PR DESCRIPTION
**Pull request by hirosassa**
Related to https://github.com/jd/tenacity/pull/315#issuecomment-871366173

In this PR, I added mypy check on CI and fixed easy type error (ignored for complicated errors, will be fixed in another PR).

### CODE REVIEW

**Review by jd — DISMISSED** [review]

**Comment by jd** [comment]
LGTM but circleci is broken again :(

**Comment by hirosassa** [comment]
Thanks for the review! I added mergify config.

> circleci is broken

Should I just wait for https://github.com/jd/tenacity/pull/340 to be merged? or Can I fix this?

**Comment by ysk24ok** [comment]
@jd Hello. Excuse me for cutting in.
What's the status of this PR? There's something I'd like to work on after this PR is merged.

**Comment by hirosassa** [comment]
@ysk24ok Thanks for the comment!
@jd Hi, jd. Could we merge this PR?

**Review by jd — APPROVED** [review]

**Comment by jd** [comment]
Not sure why tests fail?

File: tenacity/before_sleep.py [hunk 1]
```diff
@@ -36,18 +36,27 @@ def before_sleep_log(
...
+        if retry_state.outcome is None:
+            return None
+
         if retry_state.outcome.failed:
             ex = retry_state.outcome.exception()
             verb, value = "raised", f"{ex.__class__.__name__}: {ex}"
 
             if exc_info:
-                local_exc_info = retry_state.outcome.exception()
+                local_exc_info = retry_state.outcome.exception() is not None
```
Comment (ysk24ok) [inline tenacity/before_sleep.py:47]: This change breaks the existing test (`TestBeforeAfterAttempts.test_before_sleep_log_raises_with_exc_info`).
The `local_exc_info` is passed to `logger.log` so it should not be bool. 
Instead, the value should be the exception itself or None.
https://docs.python.org/3/library/logging.html#logging.LogRecord

File: tenacity/after.py [hunk 2]
```diff
@@ -36,6 +36,8 @@ def after_log(
     """After call strategy that logs to some logger the finished attempt."""
 
     def log_it(retry_state: "RetryCallState") -> None:
+        if retry_state.fn is None:
+            return None
```
Comment (ysk24ok) [inline tenacity/after.py:40]: This change breaks the existing tests (`TestAfterLogFormat.test_01_default` and `TestAfterLogFormat.test_02_custom_sec_format`).
In the real use case, `retry_state.fn` is not None (it should be the decorated function), but in tests the object is created by `make_retry_state` and the `fn` attribute is None by default.
Maybe we should manually add a dummy function to `fn` or we may use `Retrying` object to mock the decorator behavior.

File: tenacity/before_sleep.py (hunk 1 above)
Comment (hirosassa) [reply tenacity/before_sleep.py:47]: @ysk24ok Thanks for your comments! I'll take a look tonight.

**Comment by dgutson** [comment]
Hi, any update on this? @hirosassa do you need help?

**Comment by hirosassa** [comment]
@dgutson Thanks for reminding me! I'll fix this PR this weekend.

**Comment by sileht** [comment]
Thanks, @hirosassa for this, this got incorporated in https://github.com/jd/tenacity/pull/382.

**Pull request closed without merging**